src/radio.cpp

```diff
@@ -30,102 +30,140 @@ namespace bzzz
+            if(allDataFromPi.substring(0, allDataFromPi.indexOf(",")) != "S") return false;
+            // cut the channelData string after the first occurence of a comma
+            allDataFromPi = allDataFromPi.substring(allDataFromPi.indexOf(",") + 1);
 
-            for (int i = 0; i < 16; i++)
+            for (int i = 0; i < 8; i++, data_count++)
             {
                 // take the substring from the start to the first occurence of a comma,
                 // convert it to int and save it in the array
-                m_channelData[i] = allDataFromPi.substring(1, allDataFromPi.indexOf(",")).toInt();
+                // save the data to a dummy array
+                m_dummyRefData[i] = allDataFromPi.substring(0, allDataFromPi.indexOf(",")).toFloat();
 
                 // cut the channelData string after the first occurence of a comma
                 allDataFromPi = allDataFromPi.substring(allDataFromPi.indexOf(",") + 1);
             }
+            // Check if 8 data points are received and return false if not. 
+            // if yes, copy the data to the actual array
+            // this helps to retain previous data if corrupted data is received
+            if(data_count != 8) 
+            {
+                return false;
+            }
+            else 
+            { 
+                for(int i = 0; i < 8; i++) 
+                {
+                    m_refData[i] = m_dummyRefData[i];
+                }
+            }
+
+            // get the encoded switches data
+            m_dummyEncodedSwtchsData = allDataFromPi.substring(0, allDataFromPi.indexOf(",")).toInt();
+            // the max value possible for encoded switch data in binary is 0b{1 1 10 1} = 29
+            if (m_dummyEncodedSwtchsData <0 || m_dummyEncodedSwtchsData > 29)
+            {
+                return false;
+            }
+            else
+            {
+                m_encodedSwitchesData = m_dummyEncodedSwtchsData;
+            }
             return true;
         }
         return false;
     }
 
     float Radio::pitchReferenceAngleRad()
     {
-        return mapRadioToAngle(m_channelData[RADIO_CHANNEL_PITCH]);
+        return m_refData[RADIO_CHANNEL_PITCH];
     }
 
     float Radio::rollReferenceAngleRad()
     {
-        return mapRadioToAngle(m_channelData[RADIO_CHANNEL_ROLL]);
+        return m_refData[RADIO_CHANNEL_ROLL];
     }
 
     float Radio::yawRateReferenceRadSec()
     {
-        float rawRatePercentage = mapTrimmerToPercentage(m_channelData[RADIO_CHANNEL_YAW_RATE]);
-        return -RADIO_MAX_YAW_RATE_RAD_SEC + RADIO_MAX_YAW_RATE_RAD_SEC * rawRatePercentage;
+        return m_refData[RADIO_CHANNEL_YAW_RATE];
+    }
+
+    float Radio::throttleReferencePWM()
+    {
+        return m_refData[RADIO_CHANNEL_THROTTLE];
     }
 
     float Radio::throttleReferencePercentage()
     {
-        return mapTrimmerToPercentage(m_channelData[RADIO_CHANNEL_THROTTLE]);
+        return m_refData[RADIO_CHANNEL_THROTTLE]/1000 - 1;
```

Comment:
Why `-1`?

Reply:
`
        // Since PWM range is [1000, 2000], we subtract 1000 to calculate the percentage
        //   percentage = (Referance_PWM - 1000)/1000
        //=> percentage = Reference_PWM/1000 - 1
`

I have added this comment in the file.